server/src/main/java/org/elasticsearch/index/mapper/FieldMapper.java

```diff
@@ -310,6 +311,49 @@ public void parse(ParseContext context) throws IOException {
      */
     protected abstract void parseCreateField(ParseContext context) throws IOException;
 
+    /**
+     * Given access to a document's _source, return this field's values.
+     *
+     * In addition to pulling out the values, mappers can parse them into a standard form. This
+     * method delegates parsing to {@link #parseSourceValue} for parsing. Most mappers will choose
+     * to override {@link #parseSourceValue} -- for example numeric field mappers make sure to
+     * parse the  source value into a number of the right type.
+     *
+     * Some mappers may need more flexibility and can override this entire method instead.
+     *
+     * @param lookup a lookup structure over the document's source.
+     * @return a list a standardized field values.
+     */
+    public List<?> lookupValues(SourceLookup lookup) {
+        Object sourceValue = lookup.extractValue(name());
```

Comment:
Note that this PR doesn't handle the `null_value` mapping options. Currently null values are treated as if the field doesn't exist.